File: VNAE-Asymmetric-Power-Grid-Dynamics.R
# ------------------------------------------------------------
# VNAE Asymmetric Power Grid Stability
# Abstract Multi-Variable Network Model
# ------------------------------------------------------------

rm(list = ls())
set.seed(123)

# ------------------------------------------------------------
# Network parameters
# ------------------------------------------------------------

N <- 12               # number of grid nodes
T  <- 20
dt <- 0.01
time <- seq(0, T, by = dt)

# Node types:
# 1 = generator, 2 = load, 3 = renewable
node_type <- c(1,1,1, 2,2,2,2, 3,3,3,3,3)

# ------------------------------------------------------------
# Asymmetric structural parameters (VNAE core)
# ------------------------------------------------------------

# Inertia / rigidity (Theta)
theta <- c(
  1.5, 1.2, 1.0,      # generators
  0.6, 0.5, 0.7, 0.4,# loads
  0.8, 0.9, 0.6, 0.7, 0.5  # renewables
)

Theta <- diag(theta)

# Power dissipation
gamma <- diag(runif(N, 0.6, 1.2))

# ------------------------------------------------------------
# Network Laplacian (asymmetric grid)
# ------------------------------------------------------------

A <- matrix(runif(N*N, 0, 1), N, N)
A <- (A + t(A)) / 2
diag(A) <- 0

D <- diag(rowSums(A))
L <- D - A

# ------------------------------------------------------------
# Initial states
# ------------------------------------------------------------

omega <- runif(N, -0.3, 0.3)   # frequency deviations
p     <- runif(N, -1, 1)       # power injections

Omega <- matrix(0, length(time), N)
Power <- matrix(0, length(time), N)

Omega[1, ] <- omega
Power[1, ] <- p

# ------------------------------------------------------------
# Intermittent renewable forcing
# ------------------------------------------------------------

renewable_noise <- function(t, i) {
  if (node_type[i] == 3) {
    return(0.3 * sin(2 * pi * 0.4 * t) + rnorm(1, 0, 0.05))
  } else {
    return(0)
  }
}

# ------------------------------------------------------------
# Simulation loop
# ------------------------------------------------------------

for (k in 2:length(time)) {
 
  w <- Omega[k-1, ]
  p <- Power[k-1, ]
 
  # Frequency dynamics (VNAE structure)
  dw <- -L %*% w - Theta %*% w + p
 
  # Power dynamics
  dp <- -gamma %*% p
 
  for (i in 1:N) {
    dp[i] <- dp[i] + renewable_noise(time[k], i)
  }
 
  Omega[k, ] <- w + dt * dw
  Power[k, ] <- p + dt * dp
}

# ------------------------------------------------------------
# VNAE metric tensor and curvature (abstract)
# ------------------------------------------------------------

beta <- 0.2

# Metric tensor g_ij = delta_ij + beta (Theta + A)
g <- diag(N) + beta * (Theta + A)

# Scalar curvature proxy (VNAE-style)
K <- 0
for (i in 1:(N-1)) {
  for (j in (i+1):N) {
    K <- K + abs(theta[i] - theta[j]) * abs(A[i,j]) /
      (1 + beta * (theta[i] + theta[j]))
  }
}
K <- 2 * K / (N * (N - 1))

cat("VNAE scalar curvature K =", round(K, 4), "\n")

# ------------------------------------------------------------
# Visualization
# ------------------------------------------------------------

matplot(time, Omega, type = "l", lwd = 1.8,
        xlab = "Time",
        ylab = "Frequency deviation",
        main = "Asymmetric Frequency Dynamics under VNAE")
grid()

matplot(time, Power, type = "l", lwd = 1.8,
        xlab = "Time",
        ylab = "Power injection",
        main = "Power Dynamics with Renewable Intermittency")
grid()

             

File: VNAE-Asymmetric-Power-Grid-Dynamics.py
import numpy as np
import matplotlib.pyplot as plt

# ------------------------------------------------------------
# VNAE Asymmetric Power Grid Stability 
# ------------------------------------------------------------

# Set seed for reproducibility
np.random.seed(123)

# Network parameters
N = 12                # Number of grid nodes
T_max = 20.0          # Total simulation time
dt = 0.01             # Time step
time_seq = np.arange(0, T_max + dt, dt)
n_steps = len(time_seq)

# Node types: 1 = generator, 2 = load, 3 = renewable
node_type = np.array([1, 1, 1, 2, 2, 2, 2, 3, 3, 3, 3, 3])

# Asymmetric structural parameters (Theta - Rigidity/Inertia)
theta_vals = np.array([1.5, 1.2, 1.0, 0.6, 0.5, 0.7, 0.4, 0.8, 0.9, 0.6, 0.7, 0.5])
Theta = np.diag(theta_vals)

# Power dissipation parameters (Gamma)
gamma_vals = np.random.uniform(0.6, 1.2, N)
Gamma = np.diag(gamma_vals)

# Network Laplacian (L) representing grid connectivity
A = np.random.rand(N, N)
A = (A + A.T) / 2      # Symmetrize adjacency matrix
np.fill_diagonal(A, 0)
D = np.diag(np.sum(A, axis=1))
L = D - A

# Initial states (Frequency and Power)
Omega = np.zeros((n_steps, N))
Power = np.zeros((n_steps, N))
Omega[0, :] = np.random.uniform(-0.3, 0.3, N)
Power[0, :] = np.random.uniform(-1.0, 1.0, N)

# ------------------------------------------------------------
# Simulation Loop (VNAE Gradient Flow)
# ------------------------------------------------------------

for k in range(1, n_steps):
    w = Omega[k-1, :]
    p = Power[k-1, :]
    
    # Frequency dynamics: dw/dt = -Lw - Theta*w + p
    dw = -L.dot(w) - Theta.dot(w) + p
    
    # Power dynamics: dp/dt = -Gamma*p + Renewable Noise
    dp = -Gamma.dot(p)
    
    # Apply intermittent renewable forcing to type 3 nodes
    current_time = time_seq[k]
    for i in range(N):
        if node_type[i] == 3:
            # Identical noise logic: Sine wave + White Noise
            noise = 0.3 * np.sin(2 * np.pi * 0.4 * current_time) + np.random.normal(0, 0.05)
            dp[i] += noise
            
    # Euler integration step
    Omega[k, :] = w + dt * dw
    Power[k, :] = p + dt * dp

# ------------------------------------------------------------
# Visualization 
# ------------------------------------------------------------

fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(14, 6))

# Plot 1: Frequency Dynamics
for i in range(N):
    ax1.plot(time_seq, Omega[:, i], linewidth=1.2)
ax1.set_title("Asymmetric Frequency Dynamics under VNAE")
ax1.set_xlabel("Time")
ax1.set_ylabel("Frequency deviation")
ax1.grid(True, linestyle=':', alpha=0.6)

# Plot 2: Power Dynamics
for i in range(N):
    ax2.plot(time_seq, Power[:, i], linewidth=1.2)
ax2.set_title("Power Dynamics with Renewable Intermittency")
ax2.set_xlabel("Time")
ax2.set_ylabel("Power injection")
ax2.grid(True, linestyle=':', alpha=0.6)

plt.tight_layout()
plt.show()
